### PR TITLE
Add Custom Data Library Parser

### PR DESCRIPTION
**Pull request by mtc1386**
This PR adds a new data library parser to support a new YAML format. This YAML file includes more GEO Series and differs from the original YAML content as follows:

1. Added `matrix_files` field to record Series Matrix Files information.
2. Added `supplementary_files` field to record Series Supplementary Files information.
3. Added `metadata_query` in the parser to directly request geo2r API for Series Samples information.

The new YAML currently collects data from the following platforms:
```
"GPL13534": "Illumina450k",
"GPL8490": "Illumina27k",
"GPL21145": "IlluminaEPIC",
"GPL16791": "IlluminaHiSeq2500"

```

However, we encounter cases where some Series Matrix Files do not contain matrix data. The current approach is to throw an exception during load. If you have any suggestions for improvement, we would love to discuss them.

### CODE REVIEW

File: biolearn/data_library.py
Comment (sarudak) [inline biolearn/data_library.py:555]: This seems like an unnecessarily complex way of implementing this. Is there a reason we need to read the file backwards and then forwards again? Why not just find the line on a single forward pass? Then we don't need to take an additional dependency on file_read_backwards and we only need to read the file once.

File: biolearn/data_library.py
Comment (sarudak) [inline biolearn/data_library.py:563]: It would probably be better if we didn't write to disk but it we do this should be in a finally block to ensure cleanup even if an exception is thrown.

File: biolearn/data_library.py
Comment (sarudak) [inline biolearn/data_library.py:624]: If were going to put  this on the DataSource then we should use a more informative name. Data is confusing since this class gets data for us. Recommend "definition" or "source_defenition"

**Review by sarudak — CHANGES_REQUESTED** [review]
Overall looks goods with a few noted concerns. Is the new library file intentionally in the test/data folder? It seems like it should be in the data folder which is packaged with the library.

File: biolearn/data_library.py
Comment (mtc1386) [reply biolearn/data_library.py:542]: Indeed, I overthought it and used a more complex implementation. Initially, I manually checked a few large matrix files and noticed that the data was located at the bottom, so I intuitively thought reading from the back would be faster. This time, I tested both methods and found that sequential reading is faster. I have already replaced the original implementation.

File: biolearn/data_library.py
Comment (mtc1386) [reply biolearn/data_library.py]: Done.

**Comment by mtc1386** [comment]
> Overall looks goods with a few noted concerns. Is the new library file intentionally in the test/data folder? It seems like it should be in the data folder which is packaged with the library.

It's in the test/data folder just for testing. We do need to move the new library file to the data folder. I want to confirm if we should directly replace the original library.yaml?

File: biolearn/data_library.py
Comment (mtc1386) [reply biolearn/data_library.py]: Okay, I will put it in the finally block to ensure it gets cleaned up.

**Comment by sarudak** [comment]
We can't replace the existing library yaml file with this one as it doesn't contain all the datasets that are in the existing yaml file. Also my understanding is that it doesn't currently work with any datasets that don't have the methylation data in the matrix file. If that is correct we need to either prune those out for now or find a way to mark them so users don't think they're available if they're not. I think for now having it live alongside the library.yaml as it's own library file is fine. People can select it if they want to use this one and keep the existing library.yaml as the default. Once all the datasets in this file work we can merge them into one file. Let me know what you think about this solution.

**Comment by mtc1386** [comment]
> We can't replace the existing library yaml file with this one as it doesn't contain all the datasets that are in the existing yaml file. Also my understanding is that it doesn't currently work with any datasets that don't have the methylation data in the matrix file. If that is correct we need to either prune those out for now or find a way to mark them so users don't think they're available if they're not. I think for now having it live alongside the library.yaml as it's own library file is fine. People can select it if they want to use this one and keep the existing library.yaml as the default. Once all the datasets in this file work we can merge them into one file. Let me know what you think about this solution.

I agree that before we find a better way to handle datasets without methylation data, we can maintain these two library files side by side and allow users to choose between them as follows:
```
# Use default library YAML
data_library = DataLibrary()

# Use new library YAML 
library_file = get_data_file("library_extras.yaml")
data_library = DataLibrary(library_file)

```
By the way, what name do you prefer for the new library YAML? Is "library_extras.yaml" okay?

File: biolearn/data_library.py
Comment (sarudak) [inline biolearn/data_library.py:515]: This is my fault since I didn't document this but cached_download is deprecated in favor of the classes in cache.py that allow for configuration. I don't think it needs to block merge but we will need to resolve before merging the library files. Support for what you need here would require some feature additions to the cache class to support temporary file caching. No need to make any changes just a note for the future.

**Comment by sarudak** [comment]
> By the way, what name do you prefer for the new library YAML? Is "library_extras.yaml" okay?

My preference is geo_autoscan_library.yaml since it's explicit and informative about where the entries came from

**Review by sarudak — APPROVED** [review]
I think this is good to merge

**Comment by mtc1386** [comment]
> > By the way, what name do you prefer for the new library YAML? Is "library_extras.yaml" okay?
> 
> My preference is geo_autoscan_library.yaml since it's explicit and informative about where the entries came from

Done.

**Comment by sarudak** [comment]
Looking good. Couple items

- Please be sure to run the formatting
- Please delete the library file from the test files since it's in the data files. You can use the get_data_file function instead of get_test_data_file function

**Review by sarudak — APPROVED** [review]
LGTM

**Pull request merged**
